File: react/src/types/leaderboard.ts
// =====================================================
// LEADERBOARD TYPES
// =====================================================

export interface LeaderboardEntry {
  rank: number;
  playerId: string;
  username: string;
  score: number;
}

export interface GlobalLeaderboardResponse {
  periodType: string;
  periodId: string;
  entries: LeaderboardEntry[];
}

export interface WinsLeaderboardResponse {
  category: string;
  entries: LeaderboardEntry[];
}

export interface StreakLeaderboardResponse {
  category: string;
  entries: LeaderboardEntry[];
}

export interface PlayerStreakDto {
  playerId: string;
  username: string;
  currentStreak: number;
  longestStreak: number;
  lastResult: string;
}

// =====================================================
// REQUEST TYPES
// =====================================================

export interface GlobalLeaderboardRequest {
  periodType: string;
  periodId: string;
  limit?: number;
}

export interface WinsLeaderboardRequest {
  category?: string;
  limit?: number;
}

export interface StreakLeaderboardRequest {
  category?: string;
  limit?: number;
}

export type PeriodType = 'MONTHLY' | 'YEARLY' | 'ALL_TIME';

File: dotnet/NppCore/Models/Leaderboard.cs
namespace NppCore.Models;

// =====================================================
// ENTITY MODELS (mapping to Cassandra tables)
// =====================================================

/// <summary>
/// Entity for global_leaderboard table
/// </summary>
public class GlobalLeaderboardEntry
{
    public string PeriodType { get; set; } = string.Empty; // 'MONTHLY', 'YEARLY', 'ALL_TIME'
    public string PeriodId { get; set; } = string.Empty; // e.g. '2024-01'
    public int RankScore { get; set; }
    public Guid PlayerId { get; set; }
    public string Username { get; set; } = string.Empty;
}

/// <summary>
/// Entity for leaderboard_by_wins table
/// </summary>
public class WinsLeaderboardEntry
{
    public string Category { get; set; } = string.Empty; // e.g. 'most_wins'
    public int GamesWon { get; set; }
    public Guid PlayerId { get; set; }
    public string Username { get; set; } = string.Empty;
}

/// <summary>
/// Entity for player_current_streak table
/// </summary>
public class PlayerStreak
{
    public Guid PlayerId { get; set; }
    public int CurrentStreak { get; set; }
    public int LongestStreak { get; set; }
    public string LastResult { get; set; } = string.Empty; // 'WIN' or 'LOSS'
}

/// <summary>
/// Entity for leaderboard_by_longest_streak table
/// </summary>
public class StreakLeaderboardEntry
{
    public string Category { get; set; } = string.Empty; // e.g. 'global_all_time'
    public int LongestStreak { get; set; }
    public Guid PlayerId { get; set; }
    public string Username { get; set; } = string.Empty;
}

// =====================================================
// DTOs (Data Transfer Objects for API)
// =====================================================

/// <summary>
/// DTO for displaying a leaderboard entry (generic)
/// </summary>
public record LeaderboardEntryDto(
    int Rank,
    Guid PlayerId,
    string Username,
    int Score
);

/// <summary>
/// DTO for displaying streak information
/// </summary>
public record PlayerStreakDto(
    Guid PlayerId,
    int CurrentStreak,
    int LongestStreak,
    string LastResult
);

/// <summary>
/// Response for global leaderboard
/// </summary>
public record GlobalLeaderboardResponse(
    string PeriodType,
    string PeriodId,
    List<LeaderboardEntryDto> Entries
);

/// <summary>
/// Response for wins leaderboard
/// </summary>
public record WinsLeaderboardResponse(
    string Category,
    List<LeaderboardEntryDto> Entries
);

/// <summary>
/// Response for streak leaderboard
/// </summary>
public record StreakLeaderboardResponse(
    string Category,
    List<LeaderboardEntryDto> Entries
);
